source/routing/heat_map.cpp

```diff
@@ -120,60 +81,152 @@ static void print_body(std::stringstream& ss,
 }
 
 std::string print_grid(const HeatMap& heat_map) {
-    std::stringstream header;
-    std::stringstream body;
-    header << R"({"header":[)";
-    separated_by_coma(header, print_lat, heat_map.header);
-    body << R"(],"body":[)";
-    separated_by_coma(body, print_body, heat_map.body);
-    body << "]}";
-    return header.str() + body.str();
+    std::stringstream ss;
+    ss << R"({"header":[)";
+    separated_by_coma(ss, print_lat, heat_map.header);
+    ss << R"(],"body":[)";
+    separated_by_coma(ss, print_body, heat_map.body);
+    ss << "]}";
+    return ss.str();
+}
+
+static std::pair <int, int> find_rank(const BoundBox& box,
+                                       const type::GeographicalCoord& coord,
+                                       const double height_step,
+                                       const double width_step) {
+    const auto lon_rank = floor((coord.lon() - box.min.lon()) / width_step);
+    const auto lat_rank = floor((coord.lat() - box.min.lat()) / height_step);
+    return std::make_pair(lon_rank, lat_rank);
 }
 
-static HeatMap fill_heat_map(const uint step,
-                             const BoundBox& box,
-                             const double height_step,
-                             const double width_step,
-                             const georef::GeoRef& worker,
-                             const double min_dist,
-                             const double max_duration,
-                             const double speed,
-                             const std::vector<navitia::time_duration>& distances,
-                             const nt::idx_t offset) {
-    std::vector<SingleCoord> header;
-    std::vector<std::pair <SingleCoord, std::vector<navitia::time_duration>>> body;
-    for (uint i = 0; i < step; i++) {
-        std::vector<navitia::time_duration> local_duration;
-        auto lon = SingleCoord(box.min.lon() + i * width_step, width_step);
-        for (uint j = 0; j < step; j++) {
-            if (i == 0) {
-                header.push_back(SingleCoord((box.min.lat() + j * height_step), height_step));
+struct Projection {
+    boost::optional<double> distance;
+    georef::vertex_t source;
+    georef::vertex_t target;
+    Projection(double distance,
+               georef::vertex_t source,
+               georef::vertex_t target): distance(distance), source(source), target(target){}
+
+    Projection(): distance(boost::none){}
+};
+
+static std::vector<size_t> find_boundary(const std::pair <int, int>& rank_source,
```

Comment:
That should not be a vector, but a struct with names for each corners (max_y, min_y, max_x, min_x for example).